internal/terraform/node_resource_plan_instance_query.go

```diff
@@ -0,0 +1,111 @@
+	"github.com/hashicorp/terraform/internal/states"
+	"github.com/hashicorp/terraform/internal/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+func (n *NodePlannableResourceInstance) listResourceExecute(ctx EvalContext) (diags tfdiags.Diagnostics) {
+	log.Printf("[TRACE] NodePlannableResourceInstance: listing resources for %s", n.Addr)
+	config := n.Config
+	addr := n.ResourceInstanceAddr()
+	provider, providerSchema, err := getProvider(ctx, n.ResolvedProvider)
+	diags = diags.Append(err)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// validate self ref
+	diags = diags.Append(validateSelfRef(addr.Resource, config.Config, providerSchema))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	keyData := EvalDataForInstanceKey(addr.Resource.Key, nil)
+	if config.ForEach != nil {
+		forEach, _, _ := evaluateForEachExpression(config.ForEach, ctx, false)
+		keyData = EvalDataForInstanceKey(addr.Resource.Key, forEach)
+	}
+
+	// evaluate the list config block
+	var configDiags tfdiags.Diagnostics
+	configVal, _, configDiags := ctx.EvaluateBlock(config.Config, n.Schema.Body, nil, keyData)
+	diags = diags.Append(configDiags)
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// Unmark before sending to provider
+	unmarkedConfigVal, _ := configVal.UnmarkDeepWithPaths()
+	configKnown := configVal.IsWhollyKnown()
+	if !configKnown {
+		diags = diags.Append(fmt.Errorf("config is not known"))
+		return diags
+	}
+
+	log.Printf("[TRACE] NodePlannableResourceInstance: Re-validating config for %s", n.Addr)
+	validateResp := provider.ValidateListResourceConfig(
+		providers.ValidateListResourceConfigRequest{
+			TypeName: n.Config.Type,
+			Config:   unmarkedConfigVal,
+		},
+	)
+	diags = diags.Append(validateResp.Diagnostics.InConfigBody(config.Config, n.Addr.String()))
+	if diags.HasErrors() {
+		return diags
+	}
+
+	// If we get down here then our configuration is complete and we're ready
+	// to actually call the provider to list the data.
+	resp := provider.ListResource(providers.ListResourceRequest{
+		TypeName: n.Config.Type,
+		Config:   unmarkedConfigVal,
+	})
+	if resp.Diagnostics != nil {
+		return diags.Append(resp.Diagnostics.InConfigBody(config.Config, n.Addr.String()))
+	}
+
+	resources := make([]cty.Value, 0)
+	identities := make([]cty.Value, 0)
+
+	for _, evt := range resp.Results {
+		if evt.Diagnostics.HasErrors() {
+			return diags.Append(evt.Diagnostics.InConfigBody(config.Config, n.Addr.String()))
+		}
+		resources = append(resources, evt.ResourceObject)
+		identities = append(identities, evt.Identity)
+	}
+
+	vals := cty.TupleVal(resources)
+	ids := cty.TupleVal(identities)
+
+	// Create a ResourceInstanceChange for the list resource and store it in Changes
+	change := &plans.ResourceInstanceChange{
+		Addr:         n.Addr,
+		ProviderAddr: n.ResolvedProvider,
+		Change: plans.Change{
+			Action:         plans.Read,
+			Before:         cty.DynamicVal,
+			After:          vals,
+			BeforeIdentity: cty.DynamicVal,
+			AfterIdentity:  ids,
```

Comment:
I'm not sure if we should be overloading this for `list` or not. In fact I wonder if these should be combined into the list object that also contains the resource objects, that way they are directly paired and we can choose to expose those via configuration for `import`.